### ISSUE TITLE
Navigating to screen in different StackNavigator causes screen to open twice

### ISSUE THREAD

**Issue by connercms** [react-navigation/react-navigation#3265]
### Current Behavior
https://snack.expo.io/Sko4LH8Vf

When navigating to a screen in a different StackNavigator than the one currently in, the screen being opened gets put on the stack twice. In the snack above, I have created an example where a Drawer contains two items that are both themselves stack navigators. 
Steps to reproduce:
1. Click button to navigate from the Drawer Item 1 stack to a screen in Drawer Item 2 stack
2. Screen opens with back button in header
3. Pressing back button pops the screen off the stack revealing the exact same screen minus the params sent

Instead of doing `this.props.navigation.navigate('SecondScreen', { message: 'Hello'})`, I have tried supplying the name of the stack itself by doing `this.props.navigation.navigate('Stack2', { message: 'Hello'})`. This solves the issue of the screen being stacked twice but I am unable to access the params sent in the navigation action.

I have also tried the suggestion here https://github.com/react-navigation/react-navigation/issues/143 using `this.props.navigation.navigate('Stack2', {}, { type: "Navigate", routeName: "SecondScreen", params: {message: 'Hello' } } ) ` but nothing happens when I press it.

### Your Environment

| software         | version
| ---------------- | -------
| react-navigation | latest
| react-native     | .51

**Comment by 440459012** [comment]
hi, i have same problems, any news about this?

**Comment by carsonwah** [comment]
@connercms Can you try using `NavigationActions` to create the action object?

```javascript
import { NavigationActions } from 'react-navigation';
// ...
const subAction = NavigationActions.navigate({ routeName: 'SecondScreen' });
this.props.navigation.navigate('Stack2', {}, subAction);
```

I guess subAction is equivalent as below:

```javascript
{
    type:"Navigation/NAVIGATE",
    routeName:"SecondScreen"
}
```

**Comment by robsongomes** [comment]
I've used NavigationActions like you said, but the problem persists. I've noticed that componentWillReceiveProps of previous screen was called twice. Same thing with the next screen.

**Comment by connercms** [comment]
NavigationActions gives same behavior - screen opened twice. Screen on top of stack (screen that is visible) has params, screen on bottom of stack (screen visible after pressing back button) does not have params.

**Comment by purinkle** [comment]
+1

**Comment by brentvatne** [comment]
you need to specify the key, which is kind of awkward. this rfc will solve this problem: https://github.com/react-navigation/rfcs/pull/25

**Comment by jonixj** [comment]
How do I specify the key? There is not much information about this in the docs

**Comment by kosmicbird** [comment]
+1 

I'm on day 2 of trying to resolve this issue and about to go crazy. Only came up with a _very_ hackish workaround to hide it using transitionConfig but I'm not happy with hiding the problem since I lose the stack transition animations this way.

**Comment by kosmicbird** [comment]
Ok, I finally figured out how to fix my situation. I spent a nauseating amount of time on this problem, so I'll post my solution with hopes that it can help someone else in the future. 

My navigation system is deeply nested with all sorts of drawer, tab, and stack navigators within each other so I'm only posting individual sections of the system and changing my screen keys to generic names so it makes more sense for you to read. 

I was trying to navigate from **STACKCHILDKEY2** (shown below) ->

       **TABPARENTKEY1**: { screen: TabNavigator({
                            **TABCHILDKEY1**: { screen: StackNavigator({
                                       **STACKCHILDKEY1**: { screen: **FIRSTSTACKSCREEN** },
                                       **STACKCHILDKEY2**: { screen: **SECONDSTACKSCREEN** }
                                }, {
                                _// StackNavigator config options_
                                }
                            })},
                            **TABCHILDKEY2**: { screen: **NORMALCOMPONENTSCREEN** },
                        }, {
                              _// TabNavigator config options_
                            }
                        })}

From **STACKCHILDKEY2** I wanted to navigate to **OTHERSTACKCHILDKEY1**  (shown below) which is in a totally different nested nav area of my app ->

         **TABPARENTKEY2**: { screen: TabNavigator({
                            **OTHERTABCHILDKEY1**: { screen: StackNavigator({
                                     **OTHERSTACKCHILDKEY1**: { screen: **FIRSTSTACKSCREEN** },
                                     **OTHERSTACKCHILDKEY2**: { screen: **SECONDSTACKSCREEN** },
                                     **OTHERSTACKCHILDKEY3**: { screen: **THIRDSTACKSCREEN** },
                                     **OTHERSTACKCHILDKEY4**: { screen: **FOURTHSTACKSCREEN** },
                            }, {
                                _// StackNavigator config options_                             
                                }
                              )}
                          }, {
                             _// TabNavigator config options_
                            }
                          })}


The problem was that **OTHERSTACKCHILDKEY1** was stacking the screen twice (as OP mentioned) when I just wanted it to reset to the beginning of the stack without a back button duplicating the screen. In this process I was also trying to pass data around from my API which complicated it more, as I kept thinking it had something to do with how I was passing my redux data but in reality it was solely related to how I was using react-navigation. 

**MY SOLUTION**

                navigation.dispatch(NavigationActions.reset({
                    index: 0,
                    key: 'OTHERTABCHILDKEY1',
                    actions: [NavigationActions.navigate({ routeName: 'OTHERSTACKCHILDKEY1' })]
                }));


**Explanation**

I had to use NavigationActions to reset the stack I wanted to navigate to. In my case, I wanted navigate to **OTHERSTACKCHILDKEY1** (which, again, is the StackNavigator screen that was stacking twice). So I added my parent/top-level StackNavigator key **OTHERTABCHILDKEY1** to the reset "key" parameter.  In essence, I'm declaring to the reset action that I want this particular navigator to reset to initial screen in the stack. Then I added NavigationActions.navigate to indicate my desired route.

**Comment by jonixj** [comment]
I found out that this is fixed in version 2. It is in alpha but has worked well for me

**Comment by brentvatne** [comment]
some info on using key in v1: https://gist.github.com/vonovak/ef72f5efe1d36742de8968ff6a708985

**Comment by brentvatne** [comment]
resolved in 2.0
